### PR TITLE
[ADH-5240]: Implement the trino-adb connector recording function

### PR DESCRIPTION
**Pull request by VitekArkhipov**
- implemented trino-adb plugin (for writing)
- added docker compose file for starting trino with adb cluster

### CODE REVIEW

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/AdbSqlClient.java [hunk 1]
```diff
@@ -0,0 +1,1955 @@
...
+import static java.lang.Math.floorMod;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.String.format;
+import static java.sql.DatabaseMetaData.columnNoNulls;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.joining;
+
+public class AdbSqlClient
```
Comment (iamlapa) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/AdbSqlClient.java:265]: The `AdbSqlClient` class is quite large and complex, handling multiple responsibilities such as data type mapping, statistics collection, table and column management, SQL query execution, and transaction handling. This violates makes the class difficult to maintain and test.
To improve the maintainability and clarity of the code, I suggest refactoring the `AdbSqlClient` class by splitting it into smaller, logically cohesive components. Here's how we might approach this:
1. `DataTypeMapper`:
   - Issue: The class contains extensive logic for mapping data types between Trino and ADB, which adds to its complexity.
   - Solution: Extract all data type mapping logic into a separate class or package, such as `DataTypeMapper`. This class would handle methods like `toColumnMappingInternal`, `fromTrinoType`, and `toWriteMapping`.
2. `StatisticsManager`:
   - ssue: Methods related to statistics (`getTableStatistics`, `readTableStatistics`, and the nested `StatisticsDao` class) are embedded within `AdbSqlClient`.
   - Solution: Create a `StatisticsManager` class to encapsulate all statistics-related functionality.
3. `TableManager`:
   - Issue: Table and column management methods (`createTable`, `renameTable`, `setTableComment`, `setColumnComment`) increase the class's size and responsibility.
   - Solution: Move these methods into a new `TableManager` class.

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    @Provides
+    @Singleton
+    public static NodeConfig getNodeConfig(GpfdistServerConfig config, NodeManager nodeManager)
+    {
+        String internalHost = config.getServerHost() != null ? config.getServerHost() : nodeManager.getCurrentNode().getHost();
+        String externalHost = config.getServerExternalHost() != null ? config.getServerExternalHost() : internalHost;
+        return new NodeConfig()
+                .setEnvironment("adb_gpfdist")
+                .setNodeId(nodeManager.getCurrentNode().getNodeIdentifier())
+                //.setNodeBindIp(internalHost)
```
Comment (iamlapa) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java:115]: ?

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/TypeUtil.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    public static final DateTimeFormatter TIMESTAMP_TYPE_FORMATTER = new DateTimeFormatterBuilder()
+            .appendValue(ChronoField.YEAR_OF_ERA, 4, 9, SignStyle.NORMAL)
+            .parseLenient()
+            .appendPattern("-MM-dd HH:mm[:ss[.SSSSSS]][XXX][ G]")
+            .toFormatter();
+    public static final DateTimeFormatter DATE_TYPE_FORMATTER = new DateTimeFormatterBuilder()
+            .appendValue(ChronoField.YEAR_OF_ERA, 4, 9, SignStyle.NORMAL)
+            .appendPattern("-MM-dd[ G]")
+            .toFormatter();
+    public static final DateTimeFormatter TIME_TYPE_FORMATTER = DateTimeFormatter.ofPattern("HH:mm:ss.SSSSSS");
```
[the remark this answers is not in the document]
Comment (VitekArkhipov) [reply plugin/trino-adb/src/main/java/io/trino/plugin/adb/TypeUtil.java:85]: this pattern is based on trino-postgresql plugin converting types code

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/AdbSqlClient.java (hunk 1 above)
Comment (VitekArkhipov) [reply plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/AdbSqlClient.java:122]: done, except 3, cause this leads to unnecessary complexity of the code

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java (hunk 2 above)
Comment (VitekArkhipov) [reply plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java]: fixed

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        Multibinder<CreateExternalTableQueryFactory> createExtTableQueryFactories = Multibinder.newSetBinder(binder, CreateExternalTableQueryFactory.class);
+        createExtTableQueryFactories.addBinding().to(CreateReadableExternalTableQueryFactory.class).in(Scopes.SINGLETON);
+        createExtTableQueryFactories.addBinding().to(CreateWritableExternalTableQueryFactory.class).in(Scopes.SINGLETON);
+        Multibinder<InsertDataQueryFactory> insertDataQueryFactories = Multibinder.newSetBinder(binder, InsertDataQueryFactory.class);
+        insertDataQueryFactories.addBinding().to(InsertDataFromExternalTableQueryFactory.class).in(Scopes.SINGLETON);
+        insertDataQueryFactories.addBinding().to(InsertDataToExternalTableQueryFactory.class).in(Scopes.SINGLETON);
+        OptionalBinder.newOptionalBinder(binder, ConnectorPageSinkProvider.class).setBinding().to(GpfdistPageSinkProvider.class).in(Scopes.SINGLETON);
+        OptionalBinder.newOptionalBinder(binder, ConnectorRecordSetProvider.class).setBinding().to(GpfdistRecordSetProvider.class).in(Scopes.SINGLETON);
+        binder.bind(GpfdistLoadMetadataFactory.class).to(GpfdistLoadMetadataFactoryImpl.class).in(Scopes.SINGLETON);
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/GpfdistModule.java:86]: Please, use empty lines between logical blocks like in JdbcModule for readability

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/GpfdistResource.java [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            Optional<ReadContext> readContextOptional = readContextManager.get(new ContextId(tableName));
+            if (readContextOptional.isEmpty()) {
+                processNotFoundQueryRequest(tableName, asyncResponse, request);
+            }
+            else {
+                ReadContext readContext = readContextOptional.get();
+                if (initialRequest(request)) {
+                    processInitialRequest(asyncResponse, readContext, request);
+                }
+                else {
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/GpfdistResource.java:162]: nit: use else if to reduce nesting level

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/request/GpfdistReadableRequest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+{
+    private String requestId;
+    private String transactionId;
+    private String commandId;
+    private String scanId;
+    private Optional<Integer> segmentId;
+    private Optional<Integer> segmentsCount;
+    private Optional<Integer> lineDelimiterLength;
+    private short gpProtocol;
+    private Optional<String> gpMasterHost;
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/request/GpfdistReadableRequest.java:48]: do we need all this unused fields?

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/request/GpfdistWritableRequest.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import static io.trino.plugin.adb.connector.protocol.gpfdist.server.request.GpfdistRequestHeader.X_GP_XID;
+
+public class GpfdistWritableRequest
+{
+    private String requestId;
+    private String transactionId;
+    private String commandId;
+    private String scanId;
+    private Integer segmentId;
+    private Optional<Integer> segmentsCount;
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/server/request/GpfdistWritableRequest.java:40]: do we need all this unused fields?

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/load/process/GpfdistPageProcessorProvider.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public PageProcessor take()
+    {
+        lock.lock();
+        try {
+            if (!isReadyForProcessing.get()) {
+                long startTime = System.currentTimeMillis();
+                while (pageProcessors.isEmpty()) {
+                    try {
+                        if (System.currentTimeMillis() - startTime > ADB_SEGMENT_WAIT_TIMEOUT) {
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/load/process/GpfdistPageProcessorProvider.java:62]: maybe use `Supplier<Long> currentTimeMsProvider` for being able to test this class in the future?

File: plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/load/process/GpfdistPageSinkProvider.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorInsertTableHandle insertTableHandle,
+            ConnectorPageSinkId pageSinkId)
+    {
+        return createPageSinkInternal(transactionHandle, session, (ConnectorOutputTableHandle) insertTableHandle,
+                pageSinkId);
+    }
+
+    private ConnectorPageSink createPageSinkInternal(ConnectorTransactionHandle transactionHandle,
```
Comment (tigrulya-exe) [inline plugin/trino-adb/src/main/java/io/trino/plugin/adb/connector/protocol/gpfdist/load/process/GpfdistPageSinkProvider.java:116]: transactionHandle is not used

**Review by tigrulya-exe — APPROVED** [review]

**Pull request merged**